src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -793,7 +794,87 @@ public RWServerFoundException(String msg) {
             super(msg);
         }
     }
-    
+
+    static class MockableInetSocketAddress {
+        private final InetSocketAddress addr;
+
+        MockableInetSocketAddress(InetSocketAddress addr) {
+            this.addr = addr;
+        }
+
+        public String getHostName() {
+            return addr.getHostName();
+        }
+
+        public MockableInetAddress getAddress() {
+            InetAddress ia = addr.getAddress();
+            return ia == null ? null : new MockableInetAddress(ia);
+        }
+
+        @Override
+        public String toString() {
+            return addr.toString();
+        }
+    }
+
+    static class MockableInetAddress {
+        private final InetAddress ia;
+
+        MockableInetAddress(InetAddress ia) {
+            this.ia = ia;
+        }
+
+        public String getCanonicalHostName() {
+            return ia.getCanonicalHostName();
+        }
+
+        public String getHostAddress() {
+            return ia.getHostAddress();
+        }
+
+        @Override
+        public String toString() {
+            return ia.toString();
+        }
+    }
+    /**
+     * Get the name of the server principal for a SASL client.  This is visible for testing purposes.
+     * @param addr the address of the host.
+     * @param clientConfig the configuration for the client.
+     * @return the name of the principal.
+     */
+    static String getServerPrincipal(MockableInetSocketAddress addr, ZKClientConfig clientConfig) {
+        String principalUserName = clientConfig.getProperty(ZKClientConfig.ZK_SASL_CLIENT_USERNAME,
+            ZKClientConfig.ZK_SASL_CLIENT_USERNAME_DEFAULT);
+        String hostName = addr.getHostName();
+
+        boolean canonicalize = true;
+        try {
+            canonicalize = Boolean.parseBoolean(clientConfig.getProperty(ZKClientConfig.ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME,
+                ZKClientConfig.ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME_DEFAULT));
+        } catch (IllegalArgumentException ea) {
+            //ignored ...
+        }
+
+        if (canonicalize) {
+            MockableInetAddress ia = addr.getAddress();
+            if (ia == null) {
+                throw new IllegalArgumentException("Unable to canonicalize address " + addr + " because it's not resolvable");
+            }
+
+            String canonicalHostName = ia.getCanonicalHostName();
+            //avoid using literal IP address when security check fails
+            if (!canonicalHostName.equals(ia.getHostAddress())) {
```

Comment:
Would you please add test to cover this code path: canonicalization is ON, but getCanonicalHostName() returns IP address.